File: backoffice/src/shared/infrastructure/events/RabbitmqDomainEventsPublisher.py
"""
 *
 * Libraries 
 *
"""

import os
from src.shared.domain                    import DomainEvent
from src.shared.domain                    import DomainEventsPublisher
from injector                             import inject
from .JsonDomainEventSerializer           import JsonDomainEventSerializer
from amqpstorm                            import Channel
from ..eventBus.RabbitmqEventBusConnector import RabbitmqEventBusConnector
from typing                               import Any
from threading                            import Thread

"""
 *
 * Class
 *
"""

class RabbitmqDomainEventsPublisher( DomainEventsPublisher ):

    """
     *
     * Attributes 
     *
    """

    __eventBusConnector   : RabbitmqEventBusConnector
    __messageDeliveryMode : int

    """
     *
     * Methods 
     *
    """

    @inject
    def __init__( self, eventBusConnector : RabbitmqEventBusConnector ) -> object:
        self.__eventBusConnector   = eventBusConnector
        self.__messageDeliveryMode = int( os.getenv( 'EVENT_BUS_MESSAGE_DELIVERY_MODE' ) )

    def publish( self, domainEvents : list[DomainEvent] ) -> None:        
        for domainEvent in domainEvents:            
            Thread( target = self.__publishEvent, args = ( domainEvent, ) ).start()
    
    def __publishEvent( self, domainEvent : DomainEvent ) -> None:
        # Variables
        messageBody : str
        eventName   : str
        # Code
        messageBody = JsonDomainEventSerializer.serialize( domainEvent )
        eventName   = domainEvent.getEventName()
        self.__sendMessage( eventName, messageBody )
        
    def __sendMessage( self, exchangeName : str, body : str ) -> None:
        # Variables
        channel    : Channel
        properties : dict[str, Any]
        # Code
        channel    = self.__eventBusConnector.getChannel()
        properties = { 'delivery_mode' : self.__messageDeliveryMode }
        channel.basic.publish( 
            exchange    = exchangeName,
            routing_key = '',
            body        = body,
            properties  = properties
        )
        channel.close()

File: preparation/src/shared/infrastructure/events/RabbitmqDomainEventsPublisher.cs
/* 
 * 
 * Libraries
 *
*/

using System.Text;
using preparation.src.shared.domain;
using RabbitMQ.Client;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class RabbitmqDomainEventsPublisher { 

        /* 
         * 
         * Attributes
         *
        */

        private RabbitmqEventBusConnector eventBusConnector;
        private int                       messageDeliveryMode;

        /* 
         * 
         * Methods
         *
        */

        public RabbitmqDomainEventsPublisher( RabbitmqEventBusConnector eventBusConnector ) {
            this.eventBusConnector = eventBusConnector;
            messageDeliveryMode = Int32.Parse( Environment.GetEnvironmentVariable( 
                "EVENT_BUS_MESSAGE_DELIVERY_MODE"
            ) );
        }

        public async Task publish( List<DomainEvent> domainEvents ) {
            foreach( DomainEvent domainEvent in domainEvents ) {
                await Task.Run( () => {
                    publishEvent( domainEvent );
                } );
            }
        }

        private void publishEvent( DomainEvent domainEvent ) {
            // Variables
            string messageBody;
            string eventName;
            // Code
            messageBody = JsonDomainEventSerializer.serialize( domainEvent );
            eventName   = domainEvent.getEventName();
            sendMessage( eventName, messageBody );
        }

        private void sendMessage( string exchangeName, string body ) {
            // Variables
            IModel           channel;
            IBasicProperties properties;
            // code
            channel    = eventBusConnector.getChannel();
            properties = channel.CreateBasicProperties();
            properties.DeliveryMode = ( byte )messageDeliveryMode;
            channel.BasicPublish(
                exchange        : exchangeName,
                routingKey      : "", 
                body            : UTF8Encoding.UTF8.GetBytes( body ), 
                basicProperties : properties
            );
            channel.Close();
        }
    
    }

}